### PR TITLE
Implement HttpClient that automatically manages connections. 

### PR DESCRIPTION
**Pull request by kentonv**
There are actually two new client types:
- One which always connects to a given NetworkAddress, but will automatically manage a pool of reusable connections.
- One which looks up the remote address based on the URL it is given, and manages a pool of connections for each host.

The latter of these two is a "true HTTP client library".

### CODE REVIEW

File: c++/src/kj/compat/http-test.c++ [hunk 1]
```diff
@@ -2355,6 +2355,344 @@ KJ_TEST("newHttpService from HttpClient WebSockets disconnect") {
...
 // -----------------------------------------------------------------------------
 
+// TODO(now): Test NetworkAddressHttpClient:
+// - Serial requests open only one connection.
+// - Parallel requests open multiple connections.
+// - Connections time out.
+
+class CountingIoStream final: public kj::AsyncIoStream {
+  // An AsyncIoStream which waits for a promise to resolve then forwards all calls to the promised
+  // stream.
```
Comment (harrishancock) [inline c++/src/kj/compat/http-test.c++]: Inaccurate comment, looks like copy-and-paste error.

File: c++/src/kj/compat/http-test.c++ [hunk 2]
```diff
@@ -2355,6 +2355,344 @@ KJ_TEST("newHttpService from HttpClient WebSockets disconnect") {
 
 // -----------------------------------------------------------------------------
 
+// TODO(now): Test NetworkAddressHttpClient:
+// - Serial requests open only one connection.
+// - Parallel requests open multiple connections.
+// - Connections time out.
```
Comment (harrishancock) [inline c++/src/kj/compat/http-test.c++]: Remove TODO? These look implemented to me.

File: c++/src/kj/compat/http-test.c++ [hunk 3]
```diff
@@ -2355,6 +2355,344 @@ KJ_TEST("newHttpService from HttpClient WebSockets disconnect") {
...
+  KJ_EXPECT(tlsAddrCount == 1);
+
+  // Try first host again, no change in connection count.
+  doRequest(false, port1).wait(io.waitScope);
+  KJ_EXPECT(count == 2);
+  KJ_EXPECT(tlsCount == 1);
+  KJ_EXPECT(addrCount == 2);
+  KJ_EXPECT(tlsAddrCount == 1);
+
+  // Multipre requests in parallel forces more connections to that host.
```
Comment (harrishancock) [inline c++/src/kj/compat/http-test.c++]: /Multipre/Multiple/

File: c++/src/kj/compat/http-test.c++ [hunk 4]
```diff
@@ -2355,6 +2355,344 @@ KJ_TEST("newHttpService from HttpClient WebSockets disconnect") {
...
+  promise1.wait(io.waitScope);
+  promise2.wait(io.waitScope);
+  KJ_EXPECT(count == 3);
+  KJ_EXPECT(tlsCount == 1);
+  KJ_EXPECT(addrCount == 2);
+  KJ_EXPECT(tlsAddrCount == 1);
+
+  // Let everything expire.
+  clientTimer.advanceTo(clientTimer.now() + clientSettings.idleTimout * 2);
+  kj::Promise<void>(kj::NEVER_DONE).poll(io.waitScope);
```
Comment (harrishancock) [inline c++/src/kj/compat/http-test.c++]: `io.waitScope.poll()`? This is the only instance of `Promise::poll()` in these tests, so perhaps this is intentional and I just don't follow some subtle difference in their use.

Comment (kentonv) [reply c++/src/kj/compat/http-test.c++]: Oops. I implemented Promise::poll() first, then realized that I only ever used it on a NEVER_DONE promise, so added WaitScope::poll() and switched to that, but I guess I forgot to update all users.

Promise::poll() turns out not to be useful in this test, but I know I've written lots of tests in the past that could have used it, so I decided to leave it in for future use.

**Pull request merged**